pkgs/jni/src/dartjni.h

```diff
@@ -164,6 +158,33 @@ extern thread_local JNIEnv* jniEnv;
 
 extern JniContext* jni;
 
+/// Handling the lifetime of thread-local jniEnv.
+#ifndef _WIN32
+extern pthread_key_t tlsKey;
+#endif
+
+static inline void detach_thread(void* data) {
+  jniEnv = NULL;
+
+  if (*jni->jvm) {
+    (*jni->jvm)->DetachCurrentThread(jni->jvm);
+  }
+
+#ifndef _WIN32
+  pthread_key_delete(tlsKey);
```

Comment:
See other comment, TLS keys are global per process. You'd want to do this in the same place you'd free the global locks (which we currently don't do? - which may be ok given we want to keep this alive until process shuts down)